app/lib/service/secret/backend.dart

```diff
@@ -2,68 +2,126 @@
 // for details. All rights reserved. Use of this source code is governed by a
 // BSD-style license that can be found in the LICENSE file.
 
+import 'dart:convert';
+
+import 'package:clock/clock.dart';
 import 'package:gcloud/service_scope.dart' as ss;
+import 'package:googleapis/secretmanager/v1.dart' as secretmanager;
+import 'package:http/http.dart' as http;
+import 'package:logging/logging.dart';
+import 'package:pub_dev/shared/configuration.dart';
+import 'package:pub_dev/shared/monitoring.dart';
+import 'package:retry/retry.dart';
 
-import '../../shared/datastore.dart';
-import '../../shared/redis_cache.dart';
 import 'models.dart';
 
 export 'models.dart' show SecretKey;
 
+final _log = Logger('pub.secrets');
+
 /// Sets the secret backend service.
 void registerSecretBackend(SecretBackend backend) =>
     ss.register(#_secretBackend, backend);
 
 /// The active secret backend service.
 SecretBackend get secretBackend => ss.lookup(#_secretBackend) as SecretBackend;
 
-/// Represents the backend for the secret handling and related utilities.
-class SecretBackend {
-  final DatastoreDB _db;
+sealed class SecretBackend {
+  /// Lookup secret with given [id], allow an in-memory cached value no older
+  /// than [maxAge] (default 1 hour).
+  ///
+  /// Return `null`, if secret is missing, or if looking up the secret failed!
+  /// Caller is expected to gracefully degrade without the secret, or fail as it
+  /// deems fit.
+  Future<String?> lookup(
+    String id, {
+    Duration maxAge = const Duration(hours: 1),
+  });
+}
+
+final class GcpSecretBackend extends SecretBackend {
+  final http.Client _client;
+  // Map from cached secret id to:
+  // * time it was fetched, and
+  // * future secret
+  final _cache = <String, (DateTime, Future<String?>)>{};
```

Comment:
Aside: I don't mind this cache pattern, but this is yet another pattern we use for caching. May be worth to migrate/consolidate the different patterns (adding a TODO?).

Reply:
Not particularly this (hence the yet another part), and it may be worth to consolidate the other patterns to use this one.

Reply:
Yeah, the `maxAge` parameter is pretty cool.

I don't know if we want to wrap this in a nice abstraction, maybe we should do that at some point and write some pure tests for the caching functionality. It's sadly pretty popular to get caching wrong by switching less-than for greater-than or vice versus :see_no_evil: :see_no_evil: :see_no_evil: 

Even with review, and with `.before` and `.after` we frequently get caching wrong :see_no_evil: